tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/tracee/pass/TraceePassImpl.scala

```diff
@@ -0,0 +1,123 @@
+package at.forsyte.apalache.tla.tracee.pass
+
+import at.forsyte.apalache.infra.passes.DerivedPredicates
+import at.forsyte.apalache.tla.tracee._
+import at.forsyte.apalache.infra.passes.Pass.PassResult
+import at.forsyte.apalache.infra.passes.options.OptionGroup
+import at.forsyte.apalache.io.ConfigurationError
+import at.forsyte.apalache.io.json.impl.DefaultTagReader
+import at.forsyte.apalache.io.lir.TlaWriterFactory
+import at.forsyte.apalache.tla.assignments.{AssignmentOperatorIntroduction, ModuleAdapter}
+import at.forsyte.apalache.tla.imp.findBodyOf
+import at.forsyte.apalache.tla.imp.src.SourceStore
+import at.forsyte.apalache.tla.lir.transformations.TransformationTracker
+import at.forsyte.apalache.tla.lir.{ModuleProperty, NullEx, TlaModule, TlaVarDecl}
+import at.forsyte.apalache.tla.pp.NormalizedNames
+import com.google.inject.Inject
+import com.typesafe.scalalogging.LazyLogging
+
+/**
+ * The implementation of a trace evaluation encoding pass.
+ *
+ * @author
+ *   Jure Kukovec
+ */
+class TraceePassImpl @Inject() (
+    derivedPreds: DerivedPredicates.Configurable,
+    options: OptionGroup.HasTracee,
+    tracker: TransformationTracker,
+    writerFactory: TlaWriterFactory,
+    sourceStore: SourceStore)
+    extends TraceePass with LazyLogging {
+
+  override def name: String = "TraceePass"
+
+  // TODO: turn requires into proper error handling?
+  override def execute(module: TlaModule): PassResult = {
+
+    val traceSource = options.tracee.trace
+    val traceReader = new UJsonTraceReader(Some(sourceStore), DefaultTagReader)
+
+    val trace = traceReader.convert(traceReader.read(traceSource))
+    require(trace.nonEmpty)
+    if (trace.isEmpty) {
+      passFailure("The provided trace is empty.", EXITCODE_ON_EXCEPTION)
+    } else {
+
+      // TODO: make findBodyOf return Option? (unrelated to MVP)
+      val expressions = options.tracee.expressions.map { exName =>
+        findBodyOf(exName, module.operDeclarations: _*) match {
+          case NullEx => throw new ConfigurationError(s"Operator $exName undefined in module ${module.name}.")
+          case ex     => exName -> ex
+        }
+      }.toMap
+      if (expressions.isEmpty)
+        passFailure("The provided list of expressions is empty.", EXITCODE_ON_EXCEPTION)
```

Comment:
The error code name doesn't seem to fit, as this isn't stopping with an exception, unless i misunderstand. Should this not just be `ExitCodes.ERROR`?

I think i don't understand the intent of these `EXITCODE_*`  constants when we have the `ExitCodes` class already...